communication/src/AsyncTlsConnection.cpp

```diff
@@ -384,103 +385,60 @@ bool AsyncTlsConnection::verifyCertificateServer(asio::ssl::verify_context& ctx)
-                                                              std::string connectionType,
-                                                              const std::string& subject,
                                                               std::optional<NodeNum> expectedPeerId) {
-  // First, perform a basic sanity test, in order to eliminate a disk read if the certificate is
-  // unknown.
-  //
-  // The certificate must have a node id, as we put it in `OU` field on creation.
-  //
-  // Since we use pinning we must know who the remote peer is.
-  // `peerIdPrefixLength` stands for the length of 'OU=' substring
-  int peerIdPrefixLength = 3;
-  std::regex r("OU=\\d*", std::regex_constants::icase);
-  std::smatch sm;
-  regex_search(subject, sm, r);
-  if (sm.length() <= peerIdPrefixLength) {
-    LOG_WARN(logger_, "OU not found or empty: " << subject);
-    return std::make_pair(false, 0);
-  }
-
-  auto remPeer = sm.str().substr(peerIdPrefixLength, sm.str().length() - peerIdPrefixLength);
-  if (0 == remPeer.length()) {
-    LOG_WARN(logger_, "OU empty " << subject);
-    return std::make_pair(false, 0);
-  }
-
-  NodeNum remotePeerId;
-  try {
-    remotePeerId = stoul(remPeer, nullptr);
-  } catch (const std::invalid_argument& ia) {
-    LOG_WARN(logger_, "cannot convert OU, " << subject << ", " << ia.what());
-    return std::make_pair(false, 0);
-  } catch (const std::out_of_range& e) {
-    LOG_WARN(logger_, "cannot convert OU, " << subject << ", " << e.what());
-    return std::make_pair(false, 0);
-  }
-
-  // If the server has been verified, check that the peers match.
-  if (expectedPeerId) {
-    if (remotePeerId != expectedPeerId) {
-      LOG_WARN(logger_, "Peers don't match, expected: " << expectedPeerId.value() << ", received: " << remPeer);
-      return std::make_pair(false, remotePeerId);
+  uint32_t peerId = UINT32_MAX;
+  std::string conn_type;
+  bool res = concord::util::crypto::CertificateUtils::verifyCertificate(
+      receivedCert, config_.certificatesRootPath, peerId, conn_type);
+  if (expectedPeerId.has_value() && peerId != expectedPeerId.value()) return std::make_pair(false, peerId);
+  if (!res) {
+    LOG_INFO(logger_,
+             "Unable to validate certificate against the local storage, falling back to validate against the RSA "
+             "public key");
+    std::string pem_pub_key = StateControl::instance().getPeerPubKey(peerId);
+    if (!pem_pub_key.empty()) {
+      if (concord::util::crypto::Crypto::instance().getFormat(pem_pub_key) !=
+          concord::util::crypto::KeyFormat::PemFormat) {
+        pem_pub_key = concord::util::crypto::Crypto::instance()
+                          .RsaHexToPem(std::make_pair("", StateControl::instance().getPeerPubKey(peerId)))
+                          .second;
+      }
+      res = concord::util::crypto::CertificateUtils::verifyCertificate(receivedCert, pem_pub_key);
+    }
+    if (res) {
```

Comment:
Same suggestion, if you negate the condition then you can return right here and save one level of indentation. Overall, this will make it more readable.

Reply:
fixed